8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
+
+    def __repr__(self) -> str:
+        return f"{self.inst}(f={self.f})"
+
+
+class BitOpToken(vm_types.AssemblerToken):
+    """Bit-oriented file register operation token."""
+    def __init__(self, inst: str, op: int, f: str, b: str) -> None:
+        super().__init__()
+        self.inst = str(inst)
+        self.op = int(op)
+        self.f = str(f)
+        self.b = str(b)
+
+    def __repr__(self) -> str:
+        return f"{self.inst}(f={self.f}, b={self.b})"
+
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        f = resolve_int_value(assembler, self.f, 0x7f)
+        b = resolve_int_value(assembler, self.b, 0x7)
+        code = self.op | (b << 7) | f
+        return code.to_bytes(2, 'big')
+
+
+class BSFToken(BitOpToken):
+    """Bit set f token."""
+    def __init__(self, f: str, b: str) -> None:
+        super().__init__("BSF", 0x1400, f, b)
+
+
+class BCFToken(BitOpToken):
+    """Bit clear f token."""
+    def __init__(self, f: str, b: str) -> None:
+        super().__init__("BCF", 0x1000, f, b)
+
+
+class LiteralOpToken(vm_types.AssemblerToken):
+    """Literal or control operation token."""
+
+    def __init__(self, inst: str, op: int, k: str, mask: int) -> None:
+        super().__init__()
+        self.inst = str(inst)
+        self.op = int(op)
+        self.k = str(k)
+        self.mask = int(mask)
+
+    def __repr__(self) -> str:
+        return f"{self.inst}(k={self.k})"
 
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        k = resolve_int_value(assembler, self.k, self.mask)
+        code = self.op | k
+        return code.to_bytes(2, 'big')
+
+
+class GOTOToken(LiteralOpToken):
+    """Goto address token."""
+    def __init__(self, k: str) -> None:
+        super().__init__("GOTO", 0x2800, k, 0x7ff)
+
+
+class MOVLWToken(LiteralOpToken):
+    """Move literal to W token."""
+    def __init__(self, k: str) -> None:
+        super().__init__("MOVLW", 0x3000, k, 0xff)
+
+
+TEST_PROG = """
 STATUS  equ 03h
 TRISA   equ 85h
 PORTA   equ 05h
+
+        nop
         bsf STATUS,5
         movlw 00h
         movwf TRISA
         bcf STATUS,5
-Start   movlw 02h
+
+Start:  movlw 02h
         movwf PORTA
         movlw 00h
         movwf PORTA
-        goto Star
+        goto Start
 """
 
 
-class Assembler:
+class PICAssembler(vm_types.GenericAssembler):
+
+    PSEUDOOP_RE = re.compile(r"(?P<name>[A-Za-z]\w*)\s+(?P<pseudoop>\w+)\s+(?P<operands>\S+)(\s+?P<comment>;.*)?")
+    OPCODE_RE = re.compile(r"((?P<label>[A-Za-z]\w*):\s+)?(?P<opcode>\w+)(\s+(?P<operands>\S+))?(\s+?P<comment>;.*)?")
+
     def __init__(
         self,
         program_text,
         instruction_codes: type[vm_types.GenericInstructionSet],
         word_size,
     ):
         self.text = program_text
+        self.codes = instruction_codes
+        self.word_size = word_size
+        self.word_size_bytes = word_size // vm_types.BITS_IN_BYTE
+        self.symbol_table = {"map": {}, "symbols": {}}
```

Comment:
I failed to understand what `refs` was for and removed it.

`symbols` is for constants defined with `EQU` or `SET`.